YARA Engine Integration

imb0ru opened the pull request:
# YARA Engine Integration + Plugin Selection & Smart Re-execution + Extraction Controls + Authorization + UX Improvements

This PR implements a comprehensive YARA scanning integration into VolWeb, enabling malware detection and pattern matching capabilities within memory dumps for both Windows and Linux analysis, adds a plugin selection workflow after evidence upload, introduces smart re-execution with failure tracking, extraction lifecycle controls (pause/resume/stop), row-level authorization for cases and evidences, and several UX/infrastructure improvements.

## Key Features

### YARA Rule Management

- Create, import, and edit YARA rules directly in the web interface
- Import rules from `.yar` files or GitHub repositories
- Automatic syntax validation with detailed error reporting
- Status tracking with specific error codes (-1 to -4, 100 for success)
- Download individual rules as `.yar` files directly from the rule list

### YARA Ruleset Management

- Organize multiple rules into logical rulesets for grouped scanning
- Automatic ruleset compilation with binary serialization
- Validation workflow that ensures all active rules are properly compiled
- Status management for rulesets (success: 100, no active rules: -1, no valid rules: -2, compilation error: -3)
- Rules count displayed in the ruleset detail page header (number of active compiled rules)
- Download full compiled rulesets as `.yar` files

### Memory Scanning Capabilities

- Flexible scan configuration: select individual rules, entire rulesets, or custom combinations
- Integration with Volatility 3's YaraScan plugin
- Scan result persistence with "latest" scan quick access
- Detailed match information including memory offsets and matched content and rendered in the results grid
- Streaming results to JSONL: scan output is written to a per-scan JSONL file on disk instead of being held in memory, preventing OOM on large rulesets
- Paginated result fetching: frontend retrieves scan results page by page. No full result set is loaded at once
- Stop scan: dedicated stop button and endpoint send SIGTERM to the running YARA worker; previously written results are preserved
- Old result file cleanup: starting a new scan automatically deletes the previous scan's JSONL file for the same evidence
- Inline search inside scan configuration: dedicated search fields for filtering rulesets and rules when selecting what to scan, without leaving the scan dialog

### Plugin Selection Workflow

- After evidence upload, the user stays on the evidence list to monitor status in real-time via WebSocket
- Plugins are grouped by category with select-all and category-level toggles
- Required plugins (PsList, PsTree) are always included and cannot be deselected
- Optional PID filter for process-based plugins
- Optional per-plugin timeout with configurable duration
- Optional Timeliner inclusion
- Plugin search field for quick lookup by name or description
- "Reconfigure" action available from the evidence list to re-open the Plugin Selector at any time
- After starting analysis from the Plugin Selector, the user is automatically redirected back to the evidence list to monitor progress

### Extraction Lifecycle Controls (Pause / Resume / Stop)

- Pause: current plugin completes, then the engine waits — progress bar replaced by "Paused" chip
- Resume: engine continues with the next plugin — progress bar resumes
- Stop: immediately kills the running Celery worker, partial results are preserved, status set to 100
- Removed legacy "Run All" extraction, all extraction now goes through the Plugin Selector

### Evidence List Action Icons

The evidence list now provides context-aware action icons based on analysis status:

| Status | Investigate | Actions | Delete |
|--------|------------|---------|--------|
| Awaiting (status -2) | Disabled | Play — opens Plugin Selector | Active |
| Running (status 0-99, running) | Active | Pause + Stop | Disabled |
| Paused (status 0-99, paused) | Active | Resume + Stop | Disabled |
| Complete (status 100) | Active | Reconfigure — opens Plugin Selector | Active |
| Error (status -1) | Active | Reconfigure — opens Plugin Selector | Active |

### Plugin Failure Tracking

- Plugin execution errors are now captured and stored in the database (error_message field on VolatilityPlugin)
- Previously, plugin failures were silently ignored (except: pass); now they are logged and persisted
- Error handling added to all 4 wrapper plugins (Windows/Linux main and misc)
- Overview section shows 3 verbose states per plugin:
  - "Success" (green) — plugin executed and produced results
  - "Executed - No output" (grey) — plugin executed successfully but found no artifacts
  - "Failed: \<reason\>" (orange) — plugin crashed with the error message displayed
- Donut chart updated to show 3 categories: Success, Executed - No output, Failed

### Smart Re-execution

- Re-analysis from Plugin Selector now skips plugins that already completed successfully
- Only failed plugins and plugins with no output are re-executed
- New selective purge mechanism: only removes DB records for plugins being re-run, preserving successful results
- Plugin Selector shows execution status chips (Success/No output/Failed/Timed out) next to each plugin when re-configuring
- "Skip successful plugins" checkbox (enabled by default on re-runs) gives the user control
- Failed and no-output plugins are pre-selected; successful plugins are deselected by default

## Authorization & Access Control

- Row-level access control: users can only view, edit, and delete cases they are assigned to
- Evidence isolation: evidence visibility is derived from case access, users only see evidences linked to their accessible cases
- Superuser bypass: users retain full access to all cases and evidences
- Creator auto-assignment: when a non-superuser creates a case, they are automatically added to `linked_users`
- Upload session protection: `InitiateUploadView` validates case access before creating a chunked upload session

## UX Improvements

- Back navigation: ArrowBack button on EvidenceDetails page and Plugin Selector for quick return to the evidence list
- Plugin search: search field in Plugin Selector filters plugins by name or description, auto-expanding matching categories to show results
- Removed restart confirmation dialog: "Reconfigure" button navigates directly to Plugin Selector (which itself is the confirmation step via "Start Analysis")
- Investigate accessible during extraction: the Investigate button is only disabled when evidence is awaiting plugin selection, not during running/paused extraction
- Global evidence upload provider: chunked upload logic extracted from `EvidenceCreationDialog` into a global `EvidenceUploadProvider` context — upload continues in the background even if the user navigates away; a persistent snackbar shows file name and progress with a cancel button
- YARA import form lock: the evidence creation form is disabled while a YARA rule import is in progress to prevent conflicting submissions
- Fixed stuck loading spinners: `CaseDetail` and `YararulesetsDetail` now call `setLoading(false)` in a `finally` block so the spinner always stops, even on 404 responses

## Infrastructure & Bug Fixes

- Fixed CRLF line endings in `backend/entrypoint.sh` — Windows line endings caused `exec failed: no such file or directory` in Linux containers
- Auto-admin creation: `initadmin` management command added to `entrypoint.sh` — creates the default admin user automatically on first startup if it does not exist
- Removed auto-start of analysis on upload from `evidences/signals.py` — extraction tasks were being called in the `post_save` signal without being imported, causing `NameError` and 500 responses. Analysis is now triggered explicitly by the user
- Increased nginx proxy timeouts in `frontend/nginx.conf` — added proxy timeouts (600s) to prevent 504 Gateway Timeout on large file uploads
- Fixed `update_fields` on progress saves — bare `evidence.save()` in wrapper plugins was overwriting `extraction_control` back to stale values; now uses `save(update_fields=["status"])` consistently
- Fixed WebSocket stale data — added `refresh_from_db(fields=["extraction_control"])` before progress saves to prevent the `post_save` signal from broadcasting stale `extraction_control` values via WebSocket
- Suppressed jsonschema log spam — added LOGGING config to set `volatility3.schemas` logger to WARNING level
- Extraction start logging — single log line at extraction start showing plugin count and timeout configuration
- YARA scan worker memory limit — configurable `YARASCAN_MEM_LIMIT` env var (default: 2g) caps per-worker memory; workers exceeding the limit are terminated gracefully to prevent OOM conditions
- Scalable YARA scanning — scan jobs run in an isolated `yarascan` Celery queue; worker isolation prevents a single large scan from blocking other tasks
- Dashboard statistics fix — `StatisticsApiView` now correctly counts `YaraRule` objects; frontend field name corrected from `total_yara_rules` to `total_yararules`

## Performance & Scalability

- Server-side pagination for YARA rules — large rule libraries are fetched 50 at a time with DataGrid pagination; rulesets remain fully loaded
- Server-side search for rules: `SearchFilter` and `OrderingFilter` added to `YaraRuleViewSet`; the frontend sends the search query to the backend instead of filtering in memory; pagination resets automatically on new queries
- Lazy rule content loading — `YaraRuleListSerializer` excludes `rule_content`; full content is only fetched in the detail view using `defer()`
- Client-side search in ruleset list: instant filtering by name in `YaraRuleSetList` with debounced input
- Bulk-delete with single recompile — deleting multiple rules triggers a single ruleset recompilation pass instead of one per deletion; `skip_rule_validation` flag prevents redundant intermediate compilations
- WebSocket refresh strategy updated for server-side pagination consistency — partial refreshes avoid resetting the current page on live updates

imb0ru committed:
Add YARA rules and rulesets integration

Introduces backend and frontend support for YARA rule and ruleset management, validation, and scanning. Adds new Django apps for yararules and yararulesets, websocket consumers, API endpoints, Celery tasks, and Volatility engine logic for YARA operations. Updates frontend with dialogs, lists, scan components, and pages for YARA features. Improves evidence serializers, signals, and entrypoint script for better integration and permissions handling. Updates documentation to reflect YARA integration and project support.

imb0ru committed:
minor fix

imb0ru committed:
minor fixes

imb0ru committed:
frontend and version fixes

imb0ru committed:
v7 compatibility fixed

imb0ru committed:
yararule counter fixed

imb0ru committed:
rules\ruleset download feature added

imb0ru committed:
delete fab bug fixed

k1nd0ne commented:
Hi @imb0ru and thank you for this huge PR! 
That will take some time for me to review / test but I'm happy to merge it soon so it's part of the next release :)

If I have some questions I will tag you directly in the PR 👍

k1nd0ne commented on backend/volatility_engine/tasks.py:
```diff
@@ -148,3 +153,237 @@ def dump_maps(evidence_id, pid):
+    channel_layer = get_channel_layer()
+    
+    # Import here to avoid circular imports
+    from volatility_engine.engine import VolatilityEngine
```
@imb0ru I don't think this import is needed / creating circular imports ? VolatilityEngine is already imported at the top of the module and use it in many other tasks without issues it seems.

k1nd0ne commented on backend/volatility_engine/tasks.py:
```diff
@@ -148,3 +153,237 @@ def dump_maps(evidence_id, pid):
+    channel_layer = get_channel_layer()
+    
+    # Import here to avoid circular imports
+    from volatility_engine.engine import VolatilityEngine
```
@imb0ru same here :-)

k1nd0ne commented on frontend/src/components/Dialogs/YaraRuleSetCreationDialog.tsx:
```diff
@@ -0,0 +1,143 @@
+        Create a yara ruleset
+        <Typography sx={{ m: 0, pl: 2 }}>
+            <i>
+            The case will contain all of the information about your
+            investigation.
+            </i>
```
This is probably a copy/paste miss, if you can provide the appropriate description for this dialog ;-)

imb0ru committed:
refactor: remove circular import in YARA rule validation tasks
refactor: update description in YARA ruleset creation dialog

imb0ru committed:
Implement selective plugin extraction and enhance plugin management

- Added functionality to filter and select plugins based on user configuration in VolWeb.
- Introduced a new API endpoint for fetching available plugins categorized by their execution status.
- Enhanced error handling and logging for plugin execution failures.
- Updated serializers to include error messages for plugins.
- Created a new PluginSelector component for improved user interaction during evidence analysis.
- Modified existing views and tasks to support selective extraction and improved user feedback.
- Updated frontend components to display plugin execution results more clearly, including success, failure, and no output states.
- Adjusted Docker and Nginx configurations for better service management and timeout handling.

imb0ru committed:
add per-plugin timeout feature for selective extraction

imb0ru committed:
correct heading from "Linked evidences" to "Linked rules" in RulesetDetail component

imb0ru committed:
add back navigation to PluginSelector

imb0ru committed:
Add pause/resume/stop controls

imb0ru committed:
Add search functionality to PluginSelector for improved plugin management

imb0ru committed:
Add global evidence upload provider and disable form during YARA import

imb0ru committed:
Add-fix rules count to ruleset API/UI and bulk-delete with single recompile

imb0ru committed:
Add client-side search for YARA rules/rulesets

imb0ru committed:
server-side pagination, lazy rule content, and perf optimizations

imb0ru committed:
scalable YARA scan with OOM fix, stop control, and worker isolation

imb0ru committed:
Add memory limit configuration for YARA scan workers

imb0ru committed:
Implement access control for cases and evidences

imb0ru committed:
Fix loading state handling in case and ruleset detail fetching

forensicxlab commented:
Hi @imb0ru 
Thank you a lot for the collaboration on this PR ! 
Merging it now and will add some final additions before the new release this week.
